src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -195,7 +200,62 @@ internal string FullPath
             {
                 // Should already be canonicalized to a full path
                 ErrorUtilities.VerifyThrowInternalRooted(value);
-                _solutionFile = value;
+                if (FileUtilities.IsSolutionFilterFilename(value))
+                {
+                    try
+                    {
+                        using JsonDocument text = JsonDocument.Parse(File.ReadAllText(value));
+                        JsonElement solution = text.RootElement.GetProperty("solution");
+                        _solutionFile = solution.GetProperty("path").GetString();
+                        if (!File.Exists(_solutionFile))
```

Comment:
We have an abstraction for filesystem that we hope to build some caching into that should be used here:

```suggestion
                        if (!FileSystems.Default.FileExists(_solutionFile))
```